File: 0460-lfu-cache/0460-lfu-cache.cpp
class Node{
public:
    int key, value, count;
    Node *next, *prev;
    
    Node(int key, int value){
        this->key = key;
        this->value = value;
        this->count = 1;
        this->next = this->prev = NULL;
    }
};

class DLL{
private:
    Node *head, *tail;
public:
    int size;
    
    DLL(){
        this->head = new Node(-1, -1);
        this->tail = new Node(-1, -1);
        head->next = tail;
        tail->prev = head;
        this->size = 0;
    }
    
    void add(Node* node){
        head->next->prev = node;
        node->next = head->next;
        node->prev = head;
        head->next = node;
        size++;
    }
    
    void remove(Node* node){
        node->prev->next = node->next;
        node->next->prev = node->prev;
        size--;
    }
    
    Node* removeLastNode(){
        if(size == 0){
            return NULL;
        }
        Node* tailPrevNode = tail->prev;
        remove(tailPrevNode);
        return tailPrevNode;
    }
    
    void printList(){
        Node* node = head;
        while(node){
            cout<<node->key<<" ";
            node = node->next;
        }
        cout<<endl;
    }
};


class LFUCache {
private:
    unordered_map<int, Node*> nodeMap;
    unordered_map<int, DLL*> countMap;
    int capacity;
    int minFreq;
    int size;
public:
    LFUCache(int capacity) {
        this->capacity = capacity;
        this->minFreq = 1;
        this->size = 0;
    }
    
    int get(int key) {
        if(nodeMap.find(key) == nodeMap.end()){
            return -1;
        }
        updateNode(nodeMap[key]);
        return nodeMap[key]->value;
    }
    
    void put(int key, int value) {
        if(capacity == 0){
            return;
        }
        if(nodeMap.find(key) != nodeMap.end()){
            Node* node = nodeMap[key];
            node->value = value;
            updateNode(node);
        } else{
            Node* node = new Node(key, value);
            nodeMap[key] = node;
            
            if(size == capacity){
                DLL *targetList = countMap[minFreq];
                Node* lastNode = targetList->removeLastNode();
                nodeMap.erase(lastNode->key);
                size--;
            }
            
            size++;
            minFreq = 1;
            
            DLL *newList = new DLL();
            if(countMap.find(node->count) != countMap.end()){
                newList = countMap[node->count];
            }
            newList->add(node);
            countMap[node->count] = newList;
        }
    }
    
    void updateNode(Node* node){
        DLL *oldList = countMap[node->count];
        oldList->remove(node);
        if(node->count == minFreq && oldList->size == 0){
            minFreq++;
        }
        node->count++;
        DLL *newList = new DLL();
        if(countMap.find(node->count) != countMap.end()){
            newList = countMap[node->count];
        }
        newList->add(node);
        countMap[node->count] = newList;
    }
};

File: 0460-lfu-cache/0460-lfu-cache.java
class Node {
    int key;
    int value;
    int freq;
    
    Node next;
    Node prev;
    
    public Node(int key, int value, int freq) {
        this.key = key;
        this.value = value;
        
        this.freq = freq;
        
        this.next = null;
        this.prev = null;
    }
}

class DLL {
    Node head;
    Node tail;
    int size;
    
    public DLL() {
        head = new Node(-1, -1, 0);
        tail = new Node(-1, -1, 0);
        
        head.next = tail;
        tail.prev = head;
        
        size = 0;
    }
    
    public void add(Node node) {
        Node headNextNode = head.next;
        head.next = node;
        node.prev = head;
        node.next = headNextNode;
        headNextNode.prev = node;
        size++;
    } 
    
    public void remove(Node node) {
        Node nextNode = node.next;
        node.prev.next = nextNode;
        nextNode.prev = node.prev;
        size--;
    }
    
    public Node removeLastNode() {
        if(size == 0) {
            return null;
        }
        Node tailPrevNode = tail.prev;
        remove(tailPrevNode);
        return tailPrevNode;
    }
}


class LFUCache {
    Map<Integer, Node> cache;
    Map<Integer, DLL> frequencies;
    int CAPACITY;
    int minFrequence;
    
    public LFUCache(int capacity) {
        cache = new HashMap<>();
        frequencies = new HashMap<>();
        
        this.CAPACITY = capacity;
        this.minFrequence = 1;
    }
    
    public int get(int key) {
        Node node = cache.get(key);
        if(node == null) {
            return -1;
        }
        update(node);
        return node.value;
    }
    
    public void put(int key, int value) {
        Node node = cache.get(key);
        
        if(node != null) {
            node.value = value;
            update(node);
            return;
        }
        
        if(CAPACITY == cache.size()) {
            DLL head = frequencies.get(minFrequence);
            Node removedNode = head.removeLastNode();
            cache.remove(removedNode.key);
        }
        
        minFrequence = 1;
        insertNode(key, value, 1);
    }
    
    private void insertNode(int key, int value, int freq) {
        frequencies.putIfAbsent(freq, new DLL());
        
        Node newNode = new Node(key, value, freq);
        
        cache.put(key, newNode);
        frequencies.get(freq).add(newNode);
    }
    
    private void update(Node node) {
        DLL oldList = frequencies.get(node.freq);
        oldList.remove(node);
        
        if(node.freq == minFrequence && oldList.size == 0) {
            minFrequence++;
        }
        
        node.freq++; 
        
        frequencies.putIfAbsent(node.freq, new DLL());
        frequencies.get(node.freq).add(node);
    }
}